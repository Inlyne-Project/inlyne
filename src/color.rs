use std::fs::File;
use std::io::BufReader;
use std::path::PathBuf;
use std::sync::OnceLock;

use anyhow::Context;
use serde::Deserialize;
use syntect::highlighting::{
    Color as SyntectColor, Theme as SyntectTheme, ThemeSet as SyntectThemeSet,
};
use two_face::theme::EmbeddedThemeName;
use wgpu::TextureFormat;

fn hex_to_linear_rgba(c: u32) -> [f32; 4] {
    let f = |xu: u32| {
        let x = (xu & 0xff) as f32 / 255.0;
        if x > 0.04045 {
            ((x + 0.055) / 1.055).powf(2.4)
        } else {
            x / 12.92
        }
    };
    [f(c >> 16), f(c >> 8), f(c), 1.0]
}

pub fn native_color(c: u32, format: &TextureFormat) -> [f32; 4] {
    use wgpu::TextureFormat::*;
    let f = |xu: u32| (xu & 0xff) as f32 / 255.0;

    match format {
        Rgba8UnormSrgb | Bgra8UnormSrgb => hex_to_linear_rgba(c),
        _ => [f(c >> 16), f(c >> 8), f(c), 1.0],
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Theme {
    pub text_color: u32,
    pub background_color: u32,
    pub code_color: u32,
    pub quote_block_color: u32,
    pub link_color: u32,
    pub select_color: u32,
    pub checkbox_color: u32,
    pub code_highlighter: SyntectTheme,
}

impl Theme {
    pub fn dark_default() -> Self {
        static CACHED_CODE_HIGHLIGHTER: OnceLock<SyntectTheme> = OnceLock::new();
        // Initializing this is non-trivial. Cache so it only runs once
        let code_highlighter = CACHED_CODE_HIGHLIGHTER
            .get_or_init(|| ThemeDefaults::Base16OceanDark.into())
            .to_owned();
        Self {
            text_color: 0x9DACBB,
            background_color: 0x1A1D22,
            code_color: 0xB38FAC,
            quote_block_color: 0x1D2025,
            link_color: 0x4182EB,
            select_color: 0x3675CB,
            checkbox_color: 0x0A5301,
            code_highlighter,
        }
    }

    pub fn light_default() -> Self {
        static CACHED_CODE_HIGHLIGHTER: OnceLock<SyntectTheme> = OnceLock::new();
        // Initializing this is non-trivial. Cache so it only runs once
        let code_highlighter = CACHED_CODE_HIGHLIGHTER
            .get_or_init(|| ThemeDefaults::Github.into())
            .to_owned();
        Self {
            text_color: 0x000000,
            background_color: 0xFFFFFF,
            code_color: 0x95114E,
            quote_block_color: 0xEEF9FE,
            link_color: 0x5466FF,
            select_color: 0xCDE8F0,
            checkbox_color: 0x96ECAE,
            code_highlighter,
        }
    }

    pub fn code_highlighter(mut self, theme: SyntectTheme) -> Self {
        self.code_highlighter = theme;
        self
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum SyntaxTheme {
    Defaults(ThemeDefaults),
    Custom(ThemeCustom),
}

#[derive(Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct ThemeCustom {
    path: PathBuf,
}

impl SyntaxTheme {
    pub fn custom(path: PathBuf) -> Self {
        Self::Custom(ThemeCustom { path })
    }
}

impl TryFrom<SyntaxTheme> for SyntectTheme {
    type Error = anyhow::Error;

    fn try_from(syntax_theme: SyntaxTheme) -> Result<Self, Self::Error> {
        match syntax_theme {
            SyntaxTheme::Defaults(default) => Ok(SyntectTheme::from(default)),
            SyntaxTheme::Custom(ThemeCustom { path }) => {
                let mut reader = BufReader::new(File::open(&path).with_context(|| {
                    format!("Failed opening theme from path {}", path.display())
                })?);
                SyntectThemeSet::load_from_reader(&mut reader)
                    .with_context(|| format!("Failed loading theme from path {}", path.display()))
            }
        }
    }
}

// Give better error messages than regular `#[serde(untagged)]`
impl<'de> Deserialize<'de> for SyntaxTheme {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum Untagged {
            Defaults(String),
            Custom(ThemeCustom),
        }

        let Ok(untagged) = Untagged::deserialize(deserializer) else {
            return Err(serde::de::Error::custom(
                "Expects either a default theme name or a path to a custom theme. E.g.\n\
                default: \"inspired-github\"\n\
                custom:  { path = \"/path/to/custom.tmTheme\" }",
            ));
        };

        match untagged {
            // Unfortunately #[serde(untagged)] uses private internals to reuse a deserializer
            // multiple times. We can't so now we have to fall back to other means to give a good
            // error message ;-;
            Untagged::Defaults(theme_name) => match ThemeDefaults::from_kebab(&theme_name) {
                Some(theme) => Ok(Self::Defaults(theme)),
                None => {
                    let variants = ThemeDefaults::kebab_pairs()
                        .iter()
                        .map(|(kebab, _)| format!("\"{kebab}\""))
                        .collect::<Vec<_>>()
                        .join(", ");
                    let msg = format!(
                        "\"{theme_name}\" didn't match any of the expected variants: [{variants}]"
                    );
                    Err(serde::de::Error::custom(msg))
                }
            },
            Untagged::Custom(custom) => Ok(Self::Custom(custom)),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ThemeDefaults {
    Base16EightiesDark,
    Base16MochaDark,
    Base16OceanDark,
    Base16OceanLight,
    ColdarkCold,
    ColdarkDark,
    DarkNeon,
    Dracula,
    Github,
    GruvboxDark,
    GruvboxLight,
    Leet,
    MonokaiExtended,
    MonokaiExtendedLight,
    Nord,
    OneHalfDark,
    OneHalfLight,
    SolarizedDark,
    SolarizedLight,
    SublimeSnazzy,
    TwoDark,
    VisualStudioDarkPlus,
    Zenburn,
}

impl ThemeDefaults {
    fn kebab_pairs() -> &'static [(&'static str, Self)] {
        &[
            ("base16-eighties-dark", Self::Base16EightiesDark),
            ("base16-mocha-dark", Self::Base16MochaDark),
            ("base16-ocean-dark", Self::Base16OceanDark),
            ("base16-ocean-light", Self::Base16OceanLight),
            ("coldark-cold", Self::ColdarkCold),
            ("coldark-dark", Self::ColdarkDark),
            ("dark-neon", Self::DarkNeon),
            ("dracula", Self::Dracula),
            ("github", Self::Github),
            ("gruvbox-dark", Self::GruvboxDark),
            ("gruvbox-light", Self::GruvboxLight),
            ("leet", Self::Leet),
            ("monokai-extended", Self::MonokaiExtended),
            ("monokai-extended-light", Self::MonokaiExtendedLight),
            ("nord", Self::Nord),
            ("one-half-dark", Self::OneHalfDark),
            ("one-half-light", Self::OneHalfLight),
            ("solarized-dark", Self::SolarizedDark),
            ("solarized-light", Self::SolarizedLight),
            ("sublime-snazzy", Self::SublimeSnazzy),
            ("two-dark", Self::TwoDark),
            ("visual-studio-dark-plus", Self::VisualStudioDarkPlus),
            ("zenburn", Self::Zenburn),
        ]
    }

    fn from_kebab(kebab: &str) -> Option<Self> {
        Self::kebab_pairs()
            .iter()
            .find_map(|&(hay, var)| (kebab == hay).then_some(var))
    }

    pub fn as_syntect_name(self) -> &'static str {
        EmbeddedThemeName::from(self).as_name()
    }
}

impl From<ThemeDefaults> for EmbeddedThemeName {
    fn from(default: ThemeDefaults) -> Self {
        match default {
            ThemeDefaults::Base16EightiesDark => Self::Base16EightiesDark,
            ThemeDefaults::Base16MochaDark => Self::Base16MochaDark,
            ThemeDefaults::Base16OceanDark => Self::Base16OceanDark,
            ThemeDefaults::Base16OceanLight => Self::Base16OceanLight,
            ThemeDefaults::ColdarkCold => Self::ColdarkCold,
            ThemeDefaults::ColdarkDark => Self::ColdarkDark,
            ThemeDefaults::DarkNeon => Self::DarkNeon,
            ThemeDefaults::Dracula => Self::Dracula,
            ThemeDefaults::Github => Self::Github,
            ThemeDefaults::GruvboxDark => Self::GruvboxDark,
            ThemeDefaults::GruvboxLight => Self::GruvboxLight,
            ThemeDefaults::Leet => Self::Leet,
            ThemeDefaults::MonokaiExtended => Self::MonokaiExtended,
            ThemeDefaults::MonokaiExtendedLight => Self::MonokaiExtendedLight,
            ThemeDefaults::Nord => Self::Nord,
            ThemeDefaults::OneHalfDark => Self::OneHalfDark,
            ThemeDefaults::OneHalfLight => Self::OneHalfLight,
            ThemeDefaults::SolarizedDark => Self::SolarizedDark,
            ThemeDefaults::SolarizedLight => Self::SolarizedLight,
            ThemeDefaults::SublimeSnazzy => Self::SubmlimeSnazzy,
            ThemeDefaults::TwoDark => Self::TwoDark,
            ThemeDefaults::VisualStudioDarkPlus => Self::VisualStudioDarkPlus,
            ThemeDefaults::Zenburn => Self::Zenburn,
        }
    }
}

impl From<ThemeDefaults> for SyntectTheme {
    fn from(default: ThemeDefaults) -> Self {
        let default_themes = two_face::theme::extra();
        let mut theme = default_themes.get(default.into()).to_owned();

        // Github's background color is 0xfff which is the same as the default light theme
        // background. We match GitHub's website light theme code blocks instead to distinguish
        // code blocks from the background
        if default == ThemeDefaults::Github {
            theme.settings.background = Some(SyntectColor {
                r: 0xf6,
                g: 0xf8,
                b: 0xfa,
                a: u8::MAX,
            });
        }

        theme
    }
}
