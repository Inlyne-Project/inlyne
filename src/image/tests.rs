use std::path::Path;
use std::{fmt, fs};

use super::{ImageData, Px};
use crate::test_utils::init_test_log;

#[test]
fn px_parsing() {
    assert_eq!("500".parse::<Px>().unwrap(), Px(500));
    assert_eq!("500px".parse::<Px>().unwrap(), Px(500));
}

// Checks that the image crate converting to RGBA8 is the same as our technique
fn check(input_path: &Path) {
    let bytes = fs::read(input_path).unwrap();

    let expected = image::load_from_memory(&bytes)
        .unwrap()
        .into_rgba8()
        .into_vec();

    let image = ImageData::load(&bytes, false).unwrap();
    let actual = image.to_bytes();

    assert_eq!(
        Rgba8Data::new(&actual),
        Rgba8Data::new(&expected),
        "Input: {:?}",
        input_path
    );
}

#[test]
fn source_image_variety() {
    init_test_log();

    for file in [
        "rgb8.gif",
        "rgb8.jpg",
        "rgb8.png",
        "rgba8.gif",
        "rgba8.jpg",
        "rgba8.png",
    ] {
        check(&Path::new("assets").join("test_data").join(file));
    }
}

#[derive(PartialEq)]
struct Rgba8Data(Vec<[u8; 4]>);

impl Rgba8Data {
    fn new(bytes: &[u8]) -> Self {
        assert_eq!(bytes.len() % 4, 0);
        let pixels = bytes
            .chunks(4)
            .map(|chunk| match chunk {
                &[r, g, b, a] => [r, g, b, a],
                _ => unreachable!(),
            })
            .collect();
        Self(pixels)
    }
}

impl fmt::Debug for Rgba8Data {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_list()
            .entries(
                self.0
                    .iter()
                    .map(|&pixel| format!("0x{:08x}", u32::from_be_bytes(pixel))),
            )
            .finish()
    }
}
