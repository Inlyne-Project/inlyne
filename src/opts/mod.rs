mod cli;
mod config;
#[cfg(test)]
mod tests;

use std::path::{Path, PathBuf};

use crate::color;
pub use cli::{Args, ThemeType};
pub use config::{Config, FontOptions, KeybindingsSection};

use anyhow::Result;
use serde::Deserialize;
use smart_debug::SmartDebug;

#[derive(Deserialize, Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum ResolvedTheme {
    Dark,
    #[default]
    Light,
}

impl ResolvedTheme {
    fn new(theme_ty: ThemeType) -> Option<Self> {
        match theme_ty {
            ThemeType::Auto => Self::try_detect(),
            ThemeType::Dark => Some(Self::Dark),
            ThemeType::Light => Some(Self::Light),
        }
    }

    fn try_detect() -> Option<Self> {
        match dark_light::detect() {
            dark_light::Mode::Default => None,
            dark_light::Mode::Dark => Some(Self::Dark),
            dark_light::Mode::Light => Some(Self::Light),
        }
    }
}

#[derive(SmartDebug, PartialEq)]
pub struct Opts {
    pub file_path: PathBuf,
    #[debug(skip)]
    pub theme: color::Theme,
    pub scale: Option<f32>,
    pub page_width: Option<f32>,
    pub lines_to_scroll: f32,
    pub font_opts: FontOptions,
    pub keybindings: KeybindingsSection,
    pub color_scheme: Option<ResolvedTheme>,
}

impl Opts {
    pub fn parse_and_load_from(args: Args, config: Config) -> Result<Self> {
        #[cfg(test)]
        {
            // "Use" the unused params
            let (_, _) = (args, config);
            panic!("Use `Opts::parse_and_load_with_system_theme()`");
        }
        #[cfg(not(test))]
        {
            let system_color_scheme = ResolvedTheme::try_detect();
            Self::parse_and_load_inner(args, config, system_color_scheme)
        }
    }

    #[cfg(test)]
    pub fn parse_and_load_with_system_theme(
        args: Args,
        config: Config,
        theme: Option<ResolvedTheme>,
    ) -> Result<Self> {
        Self::parse_and_load_inner(args, config, theme)
    }

    fn parse_and_load_inner(
        args: Args,
        config: Config,
        fallback_theme: Option<ResolvedTheme>,
    ) -> Result<Self> {
        let Config {
            theme: config_theme,
            scale: config_scale,
            page_width: config_page_width,
            lines_to_scroll,
            light_theme,
            dark_theme,
            font_options,
            keybindings,
        } = config;

        let Args {
            file_path,
            theme: args_theme,
            scale: args_scale,
            config: _,
            page_width: args_page_width,
        } = args;

        let resolved_theme = args_theme
            .or(config_theme)
            .and_then(ResolvedTheme::new)
            .or(fallback_theme);
        let theme = {
            let (maybe_theme, fallback_values) = match resolved_theme {
                Some(ResolvedTheme::Dark) => (dark_theme, color::Theme::dark_default()),
                None | Some(ResolvedTheme::Light) => (light_theme, color::Theme::light_default()),
            };

            match maybe_theme {
                Some(theme) => theme.merge(fallback_values)?,
                None => fallback_values,
            }
        };

        let scale = args_scale.or(config_scale);
        let font_opts = font_options.unwrap_or_default();
        let page_width = args_page_width.or(config_page_width);
        let lines_to_scroll = lines_to_scroll.into();

        Ok(Self {
            file_path,
            theme,
            scale,
            page_width,
            lines_to_scroll,
            font_opts,
            keybindings,
            color_scheme: resolved_theme,
        })
    }

    /// Arguments to supply to program that are opened externally.
    pub fn program_args(file_path: &Path) -> Vec<String> {
        let current_args = Args::new();
        let mut args = Vec::new();

        args.push(file_path.display().to_string());

        if let Some(theme) = current_args.theme {
            args.push("--theme".to_owned());
            args.push(theme.as_str().to_owned());
        }

        if let Some(scale) = current_args.scale {
            args.push("--scale".to_owned());
            args.push(scale.to_string());
        }

        if let Some(config) = current_args.config {
            args.push("--config".to_owned());
            args.push(config.display().to_string());
        }

        if let Some(page_width) = current_args.page_width {
            args.push("-w".to_owned());
            args.push(page_width.to_string());
        }

        args
    }
}
