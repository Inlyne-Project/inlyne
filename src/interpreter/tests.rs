use std::collections::VecDeque;
use std::path::PathBuf;
use std::sync::{
    atomic::{AtomicU32, Ordering},
    mpsc, Arc, Mutex,
};
use std::time::{Duration, Instant};
use std::{env, thread};

use super::{HtmlInterpreter, ImageCallback, WindowInteractor};
use crate::color::{Theme, ThemeDefaults};
use crate::image::{Image, ImageData};
use crate::opts::ResolvedTheme;
use crate::test_utils::init_test_log;
use crate::utils::Align;
use crate::{Element, ImageCache};

use base64::prelude::*;
use syntect::highlighting::Theme as SyntectTheme;
use wgpu::TextureFormat;
use wiremock::{matchers, Mock, MockServer, ResponseTemplate};

// We use a dummy window with an internal counter that keeps track of when rendering a single md
// document is finished
#[derive(Clone)]
struct AtomicCounter(Arc<AtomicU32>);

impl AtomicCounter {
    fn new() -> Self {
        Self(Arc::new(AtomicU32::from(1)))
    }

    fn is_finished(&self) -> bool {
        let counter = self.0.load(Ordering::SeqCst);
        counter == 0
    }

    fn inc(&self) {
        self.0.fetch_add(1, Ordering::SeqCst);
    }

    fn dec(&self) {
        self.0.fetch_sub(1, Ordering::SeqCst);
    }
}

struct DummyWindow(AtomicCounter);

impl WindowInteractor for DummyWindow {
    fn finished_single_doc(&self) {
        self.0.dec();
    }

    fn request_redraw(&self) {}

    // The counter is inc'd for each callback we create and internally dec's when it's called
    fn image_callback(&self) -> Box<dyn ImageCallback + Send> {
        self.0.inc();
        Box::new(DummyCallback(self.0.clone()))
    }
}

struct DummyCallback(AtomicCounter);

impl ImageCallback for DummyCallback {
    fn loaded_image(&self, _: String, _: Arc<Mutex<Option<ImageData>>>) {
        self.0.dec();
    }
}

struct InterpreterOpts {
    theme: Theme,
    fail_after: Duration,
    color_scheme: Option<ResolvedTheme>,
}

impl Default for InterpreterOpts {
    fn default() -> Self {
        Self {
            theme: Theme::light_default(),
            fail_after: Duration::from_secs(8),
            color_scheme: None,
        }
    }
}

impl InterpreterOpts {
    fn new() -> Self {
        Self::default()
    }

    fn theme<T: Into<Theme>>(mut self, theme: T) -> Self {
        self.theme = theme.into();
        self
    }

    fn set_color_scheme(&mut self, color_scheme: ResolvedTheme) {
        self.color_scheme = Some(color_scheme);
    }

    fn finish(self, counter: AtomicCounter) -> (HtmlInterpreter, Arc<Mutex<VecDeque<Element>>>) {
        let Self {
            theme,
            fail_after: _,
            color_scheme,
        } = self;
        let element_queue = Arc::default();
        let surface_format = TextureFormat::Bgra8UnormSrgb;
        let hidpi_scale = 1.0;
        let file_path = PathBuf::from("does_not_exist");
        let image_cache = ImageCache::default();
        let window = Box::new(DummyWindow(counter));
        let interpreter = HtmlInterpreter::new_with_interactor(
            Arc::clone(&element_queue),
            theme,
            surface_format,
            hidpi_scale,
            file_path,
            image_cache,
            window,
            color_scheme,
        );

        (interpreter, element_queue)
    }
}

#[derive(Default)]
struct ThemeOpts {
    code_highlighter: Option<SyntectTheme>,
}

impl From<ThemeDefaults> for ThemeOpts {
    fn from(default: ThemeDefaults) -> Self {
        Self {
            code_highlighter: Some(default.into()),
        }
    }
}

impl From<ThemeOpts> for Theme {
    fn from(opts: ThemeOpts) -> Self {
        let ThemeOpts { code_highlighter } = opts;
        let mut theme = Theme::light_default();
        if let Some(code_highlighter) = code_highlighter {
            theme.code_highlighter = code_highlighter;
        }
        theme
    }
}

impl From<ThemeDefaults> for Theme {
    fn from(default: ThemeDefaults) -> Self {
        ThemeOpts::from(default).into()
    }
}

fn interpret_md(text: &str) -> VecDeque<Element> {
    interpret_md_with_opts(text, InterpreterOpts::new())
}

fn interpret_md_with_opts(text: &str, opts: InterpreterOpts) -> VecDeque<Element> {
    let fail_after = opts.fail_after;

    let counter = AtomicCounter::new();
    let (interpreter, element_queue) = opts.finish(counter.clone());
    let (md_tx, md_rx) = mpsc::channel();
    md_tx.send(text.to_owned()).unwrap();
    let interpreter_handle = std::thread::spawn(|| {
        interpreter.interpret_md(md_rx);
    });

    let start = Instant::now();
    while !counter.is_finished() {
        if interpreter_handle.is_finished() {
            panic!("The interpreter died >:V");
        } else if start.elapsed() > fail_after {
            panic!("The interpreter appeared to hang. Some task probably panicked");
        }
        thread::sleep(Duration::from_millis(1));
    }

    let mut elements_queue = element_queue.lock().unwrap();
    std::mem::take(&mut *elements_queue)
}

#[macro_export]
macro_rules! snapshot_interpreted_elements {
    ( $( ($test_name:ident, $md_text:ident) ),* $(,)? ) => {
        $crate::snapshot_interpreted_elements!(
            InterpreterOpts::new(),
            $(
                ($test_name, $md_text),
            )*
        );
    };
    ( $opts:expr, $( ($test_name:ident, $md_text:ident) ),* $(,)? ) => {
        $(
            #[test]
            fn $test_name() {
                $crate::test_utils::init_test_log();

                let text = $md_text;
                let opts = $opts;

                let htmlified = $crate::utils::markdown_to_html(
                    text,
                    opts.theme.code_highlighter.clone(),
                );
                let description = format!(" --- md\n\n{text}\n\n --- html\n\n{htmlified}");

                ::insta::with_settings!({
                    description => description,
                }, {
                    insta::assert_debug_snapshot!(interpret_md_with_opts(text, opts));
                });
            }
        )*
    }
}

#[allow(unused)]
const FOOTNOTES_LIST_PREFIX: &str = "\
This sentence[^1] has two footnotes[^2]

[^1]: 1st footnote
[^2]: 2nd footnote";

const CHECKLIST_HAS_NO_TEXT_PREFIX: &str = "\
- [x] Completed task
- [ ] Incomplete task";

const CODE_BLOCK_BG_COLOR: &str = "\
```
Fenced code block with no language tag
```

```rust
// Rust code
fn main() {}
```";

const BARE_LINK_GETS_AUTOLINKED: &str = "\
In a paragraph https://example.org

- In a list https://example.org
";

const TOML_GETS_HIGHLIGHTED: &str = "\
```toml
key = 123
```
";

const HANDLES_COMMA_IN_INFO_STR: &str = "\
```rust,ignore
let v = 1;
```
";

const UNORDERED_LIST_IN_ORDERED: &str = "\
1. 1st outer
    - bullet
2. 2nd outer
";

const NESTED_ORDERED_LIST: &str = "\
1. 1st outer
    1. 1st inner
2. 2nd outer
";

const ORDERED_LIST_IN_UNORDERED: &str = "\
- bullet
    1. 1st inner
- bullet
";

const PARA_IN_ORDERED_LIST: &str = "\
1. 1st item

    Nested paragraph

2. 2nd item
";

const CODE_IN_ORDERED_LIST: &str = "\
1. 1st item

    ```rust
    fn main() {}
    ```

2. 2nd item
";

const YAML_FRONTMATTER: &str = "\
---
date: 2018-05-01
tags:
  - another tag
---
# Markdown h1 header
";

const ALIGNED_TABLE: &str = "\
| left default | left forced | centered | right | left default |
| ------------ | :---------- | :------: | ----: | ------------ |
| text         | text        |   text   |  text | text         |
";

const UNIQUE_ANCHORS: &str = "\
# Foo
# Foo
";

// TODO: this snapshot splits `Text` up a lot more than needed
const KBD_TAG: &str = "\
Keyboard text: <kbd>Alt-\\<num\\></kbd>
";

const BLOCKQUOTE: &str = "\
> blockquote
";

const HORIZONTAL_RULER: &str = "\
horizontal ruler vv

---
";

const SMALL_TEXT: &str = "\
<small>small</small>
";

const TEXT_STYLES: &str = "\
**bold**

_italic_

~~strikethrough~~

<u>underline</u>
";

// TODO: this still has all sorts of issues (the anchor and extra whitespace)
const HEADER_INHERIT_ALIGN: &str = r##"
<div align="center">
  <h4>
    <a href="#install">
      Install
    </a>
    <span> | </span>
    <a href="#usage">
      Usage
    </a>
  </h4>
</div>"##;

const COLLAPSED_SECTION: &str = "\
<details>
<summary>summary</summary>

collapsed text
</details>
";

snapshot_interpreted_elements!(
    // (footnotes_list_prefix, FOOTNOTES_LIST_PREFIX),
    (checklist_has_no_text_prefix, CHECKLIST_HAS_NO_TEXT_PREFIX),
    (code_block_bg_color, CODE_BLOCK_BG_COLOR),
    (bare_link_gets_autolinked, BARE_LINK_GETS_AUTOLINKED),
    (toml_gets_highlighted, TOML_GETS_HIGHLIGHTED),
    (handles_comma_in_info_str, HANDLES_COMMA_IN_INFO_STR),
    (unordered_list_in_ordered, UNORDERED_LIST_IN_ORDERED),
    (nested_ordered_list, NESTED_ORDERED_LIST),
    (ordered_list_in_unordered, ORDERED_LIST_IN_UNORDERED),
    (para_in_ordered_list, PARA_IN_ORDERED_LIST),
    (code_in_ordered_list, CODE_IN_ORDERED_LIST),
    (yaml_frontmatter, YAML_FRONTMATTER),
    (aligned_table, ALIGNED_TABLE),
    (unique_anchors, UNIQUE_ANCHORS),
    (kbd_tag, KBD_TAG),
    (blockquote, BLOCKQUOTE),
    (horizontal_ruler, HORIZONTAL_RULER),
    (small_text, SMALL_TEXT),
    (text_styles, TEXT_STYLES),
    (header_inherit_align, HEADER_INHERIT_ALIGN),
    (collapsed_section, COLLAPSED_SECTION),
);

const UNDERLINE_IN_CODEBLOCK: &str = "\
```rust
use std::io;
```";

const LET_IS_ITALICIZED: &str = "\
```rust
let foo;
```";

snapshot_interpreted_elements!(
    InterpreterOpts::new().theme(ThemeDefaults::Dracula),
    (underline_in_codeblock, UNDERLINE_IN_CODEBLOCK),
    (let_is_italicized, LET_IS_ITALICIZED),
);

const NUM_IS_BOLD: &str = "\
```ts
3000;
```";

snapshot_interpreted_elements!(
    InterpreterOpts::new().theme(ThemeDefaults::Zenburn),
    (num_is_bold, NUM_IS_BOLD),
);

struct File {
    url_path: String,
    mime: String,
    bytes: Vec<u8>,
}

impl File {
    fn new(url_path: &str, mime: &str, bytes: &[u8]) -> Self {
        Self {
            url_path: url_path.to_owned(),
            mime: mime.to_owned(),
            bytes: bytes.to_owned(),
        }
    }
}

/// Spin up a server, so we can test network requests without external services
fn mock_file_server(files: &[File]) -> (MockServer, String) {
    let setup_server = async {
        let mock_server = MockServer::start().await;

        for file in files {
            let File {
                url_path,
                mime,
                bytes,
            } = file;
            Mock::given(matchers::method("GET"))
                .and(matchers::path(url_path))
                .respond_with(ResponseTemplate::new(200).set_body_raw(bytes.to_owned(), mime))
                .mount(&mock_server)
                .await;
        }

        mock_server
    };
    let server = pollster::block_on(setup_server);

    let server_url = server.uri();
    (server, server_url)
}

#[test]
fn centered_image_with_size_align_and_link() {
    init_test_log();

    let logo = include_bytes!("../../assets/test_data/bun_logo.png");
    let logo_path = "/bun_logo.png";
    let (_server, server_url) = mock_file_server(&[File::new(logo_path, "image/png", logo)]);
    let logo_url = server_url + logo_path;

    let text = format!(
        r#"
<p align="center">
  <a href="https://bun.sh"><img src="{logo_url}" alt="Logo" height=170></a>
</p>"#,
    );

    insta::with_settings!({
        // The port for the URL here is non-deterministic, but the description changing doesn't
        // invalidate the snapshot, so that's okay
        description => &text,
    }, {
        insta::assert_debug_snapshot!(interpret_md(&text));
    });
}

#[test]
fn image_loading_fails_gracefully() {
    init_test_log();

    let json = r#"{"im": "not an image"}"#;
    let json_path = "/snapshot.png";
    let (_server, server_url) =
        mock_file_server(&[File::new(json_path, "application/json", json.as_bytes())]);
    let json_url = server_url + json_path;

    let text = format!("![This actually returns JSON 😈]({json_url})");

    // Windows CI takes a very variable amount of time to handle this test specifically, and I'm
    // not sure why. Bump up the timeout delay to reduce the amount of spurious failures in as
    // tight of a niche as we can specify
    let mut opts = InterpreterOpts::new();
    if env::var("CI").map_or(false, |var| ["true", "1"].contains(&&*var.to_lowercase()))
        && cfg!(target_os = "windows")
    {
        opts.fail_after *= 2;
    }

    insta::with_settings!({
        // The port for the URL here is non-deterministic, but the description changing doesn't
        // invalidate the snapshot, so that's okay
        description => &text,
    }, {
        insta::assert_debug_snapshot!(interpret_md_with_opts(&text, opts));
    });
}

// Check to see that each paths are used for their respective color-schemes
#[test]
fn picture_dark_light() {
    fn find_image(elements: &VecDeque<Element>) -> Option<&Image> {
        elements.iter().find_map(|element| match element {
            crate::Element::Image(image) => Some(image),
            _ => None,
        })
    }

    const B64_SINGLE_PIXEL_WEBP_000: &[u8] = b"UklGRhoAAABXRUJQVlA4TA4AAAAvAAAAAM1VICICEREJAA==";
    const B64_SINGLE_PIXEL_WEBP_999: &[u8] = b"UklGRhoAAABXRUJQVlA4TA4AAAAvAAAAAM1VICICzYyIBA==";
    const B64_SINGLE_PIXEL_WEBP_FFF: &[u8] = b"UklGRhoAAABXRUJQVlA4TA4AAAAvAAAAAM1VICIC/Y+IBA==";

    init_test_log();

    let light_path = "/light.webp";
    let dark_path = "/dark.webp";
    let default_path = "/default.webp";
    let webp_mime = "image/webp";
    let files = [
        (dark_path, B64_SINGLE_PIXEL_WEBP_FFF),
        (light_path, B64_SINGLE_PIXEL_WEBP_000),
        (default_path, B64_SINGLE_PIXEL_WEBP_999),
    ]
    .map(|(path, b64_bytes)| {
        let bytes = BASE64_STANDARD.decode(b64_bytes).unwrap();
        File::new(path, webp_mime, &bytes)
    });
    let (_server, server_url) = mock_file_server(&files);
    let dark_url = format!("{server_url}{dark_path}");
    let light_url = format!("{server_url}{light_path}");
    let default_url = format!("{server_url}{default_path}");

    let text = format!(
        r#"
<p align="center">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="{dark_url}"/>
      <source media="(prefers-color-scheme: light)" srcset="{light_url}"/>
      <img src="{default_url}"/>
    </picture>
</p>
"#,
    );

    for color_scheme in [None, Some(ResolvedTheme::Dark), Some(ResolvedTheme::Light)] {
        let mut opts = InterpreterOpts::new();
        if let Some(color_scheme) = color_scheme {
            opts.set_color_scheme(color_scheme);
        }
        let elements = interpret_md_with_opts(&text, opts);
        let image = find_image(&elements).unwrap();

        // Should pick up align from the enclosing `<p>`
        assert_eq!(image.is_aligned, Some(Align::Center));

        let rgba_data = image
            .image_data
            .lock()
            .unwrap()
            .as_ref()
            .unwrap()
            .to_bytes();
        let byte = match color_scheme {
            Some(ResolvedTheme::Dark) => 0xff,
            Some(ResolvedTheme::Light) => 0x00,
            None => 0x99,
        };
        // Image data stores pixel data in RGBA format
        let expected = [byte, byte, byte, 0xff];
        assert_eq!(
            rgba_data, expected,
            "Failed for color scheme: {color_scheme:?}"
        );
    }
}
