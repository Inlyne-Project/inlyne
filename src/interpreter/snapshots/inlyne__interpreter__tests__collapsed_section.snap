---
source: src/interpreter/tests.rs
description: " --- md\n\n<details>\n\n<summary>summary</summary>\n\ncollapsed text\n\n</details>\n\n\n --- html\n\n<details>\n<summary>summary</summary>\n<p>collapsed text</p>\n</details>\n"
expression: "interpret_md_with_opts(text, opts)"
---
[
    Spacer(
        InvisibleSpacer(5),
    ),
    Spacer(
        InvisibleSpacer(5),
    ),
    Section(
        Section {
            elements: [
                Positioned {
                    inner: TextBox(
                        TextBox {
                            texts: [
                                Text {
                                    text: "collapsed text",
                                    default_color: Color(BLACK),
                                    ..
                                },
                            ],
                            ..
                        },
                    ),
                    bounds: None,
                },
            ],
            hidpi_scale: 1.0,
            hidden: RefCell {
                value: true,
            },
            summary: Some(
                Positioned {
                    inner: TextBox(
                        TextBox {
                            texts: [
                                Text {
                                    text: "summary",
                                    default_color: Color(BLACK),
                                    ..
                                },
                            ],
                            ..
                        },
                    ),
                    bounds: None,
                },
            ),
        },
    ),
    Spacer(
        InvisibleSpacer(5),
    ),
]
